src/org/rascalmpl/util/maven/Dependency.java

```diff
@@ -0,0 +1,191 @@
+        }
+
+        var exclusions = d.getExclusions().stream()
+            .map(e -> e.getGroupId() + ":" + e.getArtifactId())
+            .collect(Collectors.toUnmodifiableList());
+
+
+        var existing = context.dependencyCache.get(coodinate);
+        if (existing != null && existing.exclusions.equals(exclusions)) {
+            // do not calculate it again, but reuse for performance (some deps are very common)
+            if (existing.scope == scope) {
+                return existing;
+            }
+            return new Dependency(existing.coordinate, scope, existing.found, existing.optional, existing.dependencies, existing.exclusions);
+        }
+
+        if (context.cycleDetection.contains(coodinate)) {
+            // we're in a cycle, so lets break it by returning a version of ourselves without calculated dependencies (and don't put it in the cache)
+            return new Dependency(coodinate, scope, false, d.isOptional(), Collections.emptyList(), exclusions);
+        }
+        context.cycleDetection.add(coodinate);
+        try {
+            var dependencies = buildDependencies(d, messages, context);
+            var result = new Dependency(coodinate, scope, dependencies != null, d.isOptional(), dependencies == null ? Collections.emptyList() : dependencies, exclusions);
+            if (existing != null) {
+                context.dependencyCache.put(coodinate, result);
+            }
+            return result;
+        } finally {
+            context.cycleDetection.remove(coodinate);
+        }
+    }
+
+    private static @Nullable List<Dependency> buildDependencies(org.apache.maven.model.Dependency me, IListWriter messages,
+        CurrentResolution context) {
+        try {
+            if (me.getVersion() == null) {
+                // TODO: figure out how maven resolves a dependency that has no version number and see if we want to support this
+                throw new UnresolvableModelException("Null version not supported right now", me.getGroupId(), me.getArtifactId(), me.getVersion());
+            }
+            var resolvedEntry = context.resolver.resolveModel(me);
+            var fullDependencies = Project.parseRepositoryPom(resolvedEntry, context).getDependencies();
+            var exclusions = me.getExclusions();
+            if (exclusions.isEmpty()) {
+                return fullDependencies;
+            }
+            return fullDependencies.stream()
+                .filter(d -> !d.matches(exclusions))
+                .collect(Collectors.toUnmodifiableList());
+        } catch (UnresolvableModelException e) {
+            var loc = context.pom;
+            var artifactLoc = me.getLocation("artifactId");
+            if (artifactLoc != null) {
+                loc = IRascalValueFactory.getInstance().sourceLocation(loc , 0, 0, artifactLoc.getLineNumber(), artifactLoc.getColumnNumber(), artifactLoc.getLineNumber(), artifactLoc.getColumnNumber() + 1);
```

Comment:
This is broken. Source locations with 0,0 as offset and length are never queried for line and column information. The two have to align; as they are intended to be redundant information.